internal/stores/objects.go

```diff
@@ -465,15 +455,47 @@ func (s *SQLStore) Put(key string, o object.Object) error {
 			}
 
 			for _, shard := range ss.Shards {
-				// Create shard.
-				err = tx.Create(&dbSector{
-					SlabID: slab.ID,
-					Host:   shard.Host,
-					Root:   shard.Root,
+				// Translate pubkey to contract.
+				fcid, exists := usedContracts[shard.Host]
+				if !exists {
+					return fmt.Errorf("error converting host pubkey %v to contract id", shard.Host)
+				}
+
+				// Create sector if it doesn't exist yet.
+				var sector dbSector
+				err = tx.FirstOrCreate(&sector, &dbSector{
+					Root: shard.Root,
 				}).Error
 				if err != nil {
 					return err
 				}
+
+				// Append the sector to the slab.
+				err = tx.Model(&slab).
+					Association("Shards").
+					Append(&sector)
+				if err != nil {
+					return err
+				}
+
+				// Look for the contract referenced by the shard.
+				var contract dbContractRHPv2
+				err = tx.Model(&dbContractRHPv2{}).
+					Where(&dbContractRHPv2{FCID: fcid}).
+					Take(&contract).Error
+				if errors.Is(err, gorm.ErrRecordNotFound) {
+					continue // don't set contract
```

Comment:
can we `continue` here just like that?

Reply:
In theory we can. It just means that the sector is unhealthy from the start. We can return an error here as well which would be a little less graceful since it causes the whole upload to fail if a single contract was deleted/renewed in the meantime.

Reply:
For safety I'd keep expired and renewed contracts in a separate table. Otherwise we need to be aware of and ignore expired/renewed contracts in 99% of queries that we write and we also end up querying an ever-growing table.

Reply:
we could also have an `active` column with an index, it's not that much cognitive load to query for `active` contracts and the fact that all contracts are in one table will have some benefits I think, including this case here but I was kind of hoping to be able to build recursive queries to fetch a contract's history in a single query